tensorflow/contrib/go/graph.go

```diff
@@ -0,0 +1,553 @@
+package tensorflow
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"strings"
+
+	"github.com/golang/protobuf/proto"
+
+	pb "github.com/tensorflow/tensorflow/tensorflow/contrib/go/proto"
+)
+
+const (
+	cOpsProtobufDefsPath = "/usr/local/tensorlow/ops.pbtxt"
+)
+
+// A Graph is the representation of the computation graph.
+type Graph struct {
+	def *pb.GraphDef
+
+	availableOps map[string]*pb.OpDef
+	constants    map[string]*Tensor
+	variables    map[string]*Tensor
+}
+
+// A GraphNode is the representation of one of the nodes of the TensorFlow
+// Graph. A node takes zero or more Tensors, performs some computation, and
+// produces zero or more Tensors.
+type GraphNode struct {
+	ref          *pb.NodeDef
+	def          *pb.NodeDef
+	outDataTypes map[string]DataType
+}
+
+// NewGraph returns an initialized instance of the Graph struct.
+func NewGraph() *Graph {
+	return &Graph{
+		def:          new(pb.GraphDef),
+		availableOps: make(map[string]*pb.OpDef),
+		constants:    make(map[string]*Tensor),
+		variables:    make(map[string]*Tensor),
+	}
+}
+
+// NewGraphFromReader reads from reader until an error or EOF and loads the
+// content into a new graph. Use the asText parameter to specify if the graph
+// from the reader is provided in Text format.
+func NewGraphFromReader(reader io.Reader, asText bool) (gr *Graph, err error) {
```

Comment:
I added asText based on the [write_graph](https://www.tensorflow.org/versions/r0.8/api_docs/python/train.html#write_graph) python implementation, but in the [import_graph_def](https://www.tensorflow.org/versions/r0.8/api_docs/python/framework.html#import_graph_def) the as_text is not used so I suppose that they are doing as you say, they use auto-detection.
The problem if try to unmarshal as text and if it fails using proto.Unmarshal directly is worst in terms of performance and I wouldn't know what error message to return in case of an error, but in the other hand could be more easy to understand for the developers.

what are your thoughts on this? @tmv @dave-andersen @vrv